073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,250 @@
+Including the state in the item name facilitates understanding the current role of the Kubernetes Secret within the cluster.
+
+Problems from multiple processes concurrently updating the Kubernetes Secret are avoided by using `metadata.resourceVersion` to make conditional requests when updating that Secret. Note also that only one operator makes each state transition.
+
+### CA certificate state machine
+
+The "CA certificate trust state machine" has the following states:
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_UNUSED // trusted everywhere it needs to be, but no EE certificates yet issued with this CA certificate
+   |
+   v
+  TRUSTED_IN_USE // trusted everywhere, EE certificates issued using this CA certificate
+   |
+   v
+  PHASE_OUT // trusted everywhere, but new end-entity certificates will not be issued using this CA certificate
+   |
+   v
+  ZOMBIE // still trusted by some component, but with no issued end-entity certificates in current use.
+```
+
+When a new cluster is first created the flow is:
+
+1. Either user or CaReconciler creates the Cluster and Client CA certificates and keys and stores them in Kubernetes Secrets.
+2. The CaReconciler takes the Cluster CA certificate and places it in the new shared Kubernetes Secret with the state UNTRUSTED
+3. At the beginning of the reconcile loop of components, the cluster operator generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret
+   It does this even though the state is UNTRUSTED, because the Kubernetes Secret was missing
+4. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc):
+   1. generate their certificates, store them in Kubernetes Secrets
+   2. create their pods annotated with the certificates Kubernetes Secret and new shared Kubernetes Secret mounted onto the pod
+5. At the end of the reconcile loop of components, the cluster operator updates the shared Kubernetes Secret to change the state of the CA certificate to TRUSTED_IN_USE
+
+When the CA private key is updated (either because it has expired or the user has incremented the generation to indicate they've manually updated the Kubernetes Secret):
+
+1. If Strimzi is managing the CA, the CaReconciler renews the Cluster and Client CAs certificates and keys. The old certificate and key can be removed from the Kubernetes Secret.
+2. The CaReconciler takes the Cluster CA certificate and places it in the new shared Kubernetes Secret with the state UNTRUSTED. It also updates the old CA certificate to have the state PHASE_OUT.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) see the new UNTRUSTED CA certificate and:
+   1. Roll the pods to trust the new CA certificate
+4. At the end of the reconcile loop the cluster operator updates the shared Kubernetes Secret to change the state of the CA certificate to TRUSTED_UNUSED
```

Comment:
Again, you might want to track it in some secret. But as long as we need to roll the pods, you need to track it in the pods in the first place.